llvm/lib/Transforms/IPO/SampleProfileMatcher.cpp

```diff
@@ -782,6 +782,25 @@ bool SampleProfileMatcher::functionMatchesProfileHelper(
   float Similarity = 0.0;
 
   const auto *FSFlattened = getFlattenedSamplesFor(ProfFunc);
+  // With extbinary profile format, initial profile loading only reads profile
+  // based on current function names in the module.
+  // However, if a function is renamed, sample loader fails to load its original
+  // profile(which has a different name), we will miss this case. To address
+  // this, we load the top-level profile candidate explicitly for the matching.
+  if (!FSFlattened) {
+    DenseSet<StringRef> TopLevelFunc({ProfFunc.stringRef()});
+    SampleProfileMap TopLevelProfile;
+    Reader.read(TopLevelFunc, TopLevelProfile);
```

Comment:
If we populate the existing Profiles map in reader, the `read` API won't need a `SampleProfileMap` parameter.

Reply:
We can still make the API simpler for reader - we can always load into reader's profile map, and then retriever it by name for downstream use. Reader can serve as a cache too.

OTOH, the profile from reader isn't guaranteed to be flattened, why do we put it directly into flattened profile map anyways?

Reply:
Sounds good. It's a little bit tricky, there could be more places to use the `FlattenedProfiles`(the profiles for the matching), then anywhere it used, we also need to original samples reader.  (I was thinking to rename `FlattenedProfiles` and use it so that we only have one container.). Right now there is only one more place, just updated that.